### ISSUE TITLE
Caddy tries to get certificate for domains only in host matcher within site block

### ISSUE THREAD

**Issue by lxhao61** [caddyserver/caddy#5124]
Does JSON deactivate [host matcher](https://caddyserver.com/docs/json/apps/http/servers/routes/match/host/) to activate automatic HTTPS？
Example:
```
{
	order forward_proxy before file_server
	admin off
	log {
		output file /var/log/caddy/access.log
		level ERROR
	}
	storage file_system {
		root /home/tls
	}

	auto_https off
	default_sni *.xx.yy
	cert_issuer zerossl
	acme_dns duckdns b8a33cc1-1111-2222-3333-b049bf07988f {
		override_domain zz.duckdns.org
	}

	servers unix//dev/shm/h1h2c.sock {
		listener_wrappers {
			proxy_protocol
		}
		protocols h1 h2c
	}

	servers 127.0.0.1:8443 {
		listener_wrappers {
			proxy_protocol
			tls
		}
	}
}

:80 {
	redir https://{host}{uri} permanent
}

:88 {
	bind unix//dev/shm/h1h2c.sock

	file_server {
		root /var/www/html
	}
}

:8443, *.xx.yy:8443 {
	bind 127.0.0.1

	forward_proxy {
		basic_auth user pass
		hide_ip
		hide_via
		probe_resistance
	}

	@host2 {
		host wh.xx.yy ch.xx.yy
	}
	file_server @host2 {
		root /var/www/html
	}
}
```
The above example has been normal before. 
After [this master update](https://github.com/caddyserver/caddy/commit/2808de1e30b873f9c8f4693eae59bd56f20ebbbe), the automatic parameter is lost in the TLS configuration, that is, the certificate cannot be applied for *.xx.yy. 
If the above configuration deletes `auto_https off`, it is found that the background program automatically applies for a certificate for wh.xx.yy、ch.xx.yy.
Based on the above example, thinking about the problem of the update: I think it is the most reasonable method to cancel the host matcher to activate automatic HTTPS. Benefit: Makes the Caddyfile well-configured. Make the host matcher flexible.

**Comment by francislavoie** [comment]
If you set `auto_https off`, then you're telling Caddy "don't manage certificates for me". So don't set that setting if you _do_ want managed certificates.

I'm not sure I understand the intent of your config. It doesn't make sense to me what you're trying to do.

I strongly recommend removing that `tls` config by the way; let Caddy use its defaults for security parameters, so that when you upgrade, you get the best modern security.

**Comment by lxhao61** [comment]
My TLS requirements are: apply for a certificate for *.xx.yy, but not for wh.xx.yy and ch.xx.yy. The purpose of configuring auto_https off in the above example is to prohibit applying for certificates for wh.xx.yy and ch.xx.yy.
v2.6.1之前版本（Caddy转JSON）：
```
.......
          {
            "match": [{
              "host": ["wh.xx.yy","ch.xx.yy"]
            }],
            "handle": [{
              "handler": "file_server",
              "root": "/var/www/html"
            }]
          }],
          "tls_connection_policies": [{
            "default_sni": "*.itdiy.gq"
          }],
          "automatic_https": {
            "disable": true
          }
        }
      }
    },
    "tls": {
      "certificates": {
        "automate": ["*.xx.yy"]
      },
      "automation": {
        "policies": [{
          "issuers": [{
            "module": "zerossl",
            "challenges": {
              "dns": {
                "provider": {
                  "name": "duckdns",
                  "api_token": "b8a33cc1-1111-2222-3333-b049bf07988f",
                  "override_domain": "zz.duckdns.org"
                }
              }
            }
          }]
        }]
      }
    }
  }
}
```
After [this master is updated](https://github.com/caddyserver/caddy/commit/2808de1e30b873f9c8f4693eae59bd56f20ebbbe) (Caddy to JSON)：
```
.......
          {
            "match": [{
              "host": ["wh.xx.yy","ch.xx.yy"]
            }],
            "handle": [{
              "handler": "file_server",
              "root": "/var/www/html"
            }]
          }],
          "tls_connection_policies": [{
            "default_sni": "*.itdiy.gq"
          }],
          "automatic_https": {
            "disable": true
          }
        }
      }
    },
    "tls": {
      "automation": {
        "policies": [{
          "issuers": [{
            "module": "zerossl",
            "challenges": {
              "dns": {
                "provider": {
                  "name": "duckdns",
                  "api_token": "b8a33cc1-1111-2222-3333-b049bf07988f",
                  "override_domain": "zz.duckdns.org"
                }
              }
            }
          }]
        }]
      }
    }
  }
}
```
After [this master is updated](https://github.com/caddyserver/caddy/commit/2808de1e30b873f9c8f4693eae59bd56f20ebbbe) , delete auto_https off (Caddy to JSON)：
```
.......
          {
            "match": [{
              "host": ["wh.xx.yy","ch.xx.yy"]
            }],
            "handle": [{
              "handler": "file_server",
              "root": "/var/www/html"
            }]
          }],
          "tls_connection_policies": [{
            "default_sni": "*.itdiy.gq"
          }]
        }
      }
    },
    "tls": {
      "certificates": {
        "automate": ["*.xx.yy"]
      },
      "automation": {
        "policies": [{
          "issuers": [{
            "module": "zerossl",
            "challenges": {
              "dns": {
                "provider": {
                  "name": "duckdns",
                  "api_token": "b8a33cc1-1111-2222-3333-b049bf07988f",
                  "override_domain": "zz.duckdns.org"
                }
              }
            }
          }]
        }]
      }
    }
  }
}
```

**Comment by francislavoie** [comment]
> My TLS requirements are: apply for a certificate for *.xx.yy, but not for wh.xx.yy and ch.xx.yy.

I don't understand what you're expecting to happen. Caddy _needs_ a TLS certificate for HTTPS. How could it have solve TLS handshakes for wildcard but _not_ those two subdomains?

Why do you need to turn off auto_https? What are you trying to prevent with this? All that does in your case is turn off the HTTP->HTTPS redirects, I think, in your old config. If that's what you want, then use `auto_https disable_redirects`. It's wrong to turn it off completely for that usecase.

**Comment by lxhao61** [comment]
> > My TLS requirements are: apply for a certificate for *.xx.yy, but not for wh.xx.yy and ch.xx.yy.
> 
> I don't understand what you're expecting to happen. Caddy _needs_ a TLS certificate for HTTPS. How could it have solve TLS handshakes for wildcard but _not_ those two subdomains? There's a logic error here.
> 
> Why do you need to turn off auto_https? What are you trying to prevent with this? All that does in your case is turn off the HTTP->HTTPS redirects, I think, in your old config. If that's what you want, then use `auto_https disable_redirects`. It's wrong to turn it off completely for that usecase.

I expect to request a certificate only for wildcard domains, not for subdomains. I use SNI for domain name traffic.
I use auto_https off to disable certificate requests for subdomains.
I have tested to use auto_https disable_redirects, and found that the background is still applying for a certificate for the subdomain.
So how to solve this problem? Applying for a certificate for a subdomain is prohibited.
In addition, I update my description above, I hope the expression is accurate. The above is my actual test, I hope you will understand after trying it.

**Comment by mholt** [comment]
Ok, well, saying `auto_https off` disables all automatic HTTPS including certificate automation, so it won't make a wildcard certificate for you.

I think the correct thing to do would be to remove `auto_https off`.

@francislavoie I did just notice, however, that his host matcher for `wh.itdiy.gq ch.itdiy.gq` is in a subroute, yet the adapted JSON has it at a top-level route:

```
  "apps": {
    "http": {
      "servers": {
        "srv0": {
          "listen": [
            "127.0.0.1:8443"
          ],
          "routes": [
            {
              "match": [
                {
                  "host": [
                    "wh.xx.yy",
                    "ch.xx.yy"
                  ]
                }
              ],
              "handle": [
                {
                  "handler": "subroute",
                  "routes": [
                    {
                      "handle": [
                        {
                          "handler": "headers",
...
```

which seems like a bug. `host` headers used within a site block should be within a subroute so auto-HTTPS isn't applied to it.

I can't look into this for a little while because of my current backlog and some recent family things, but anyone else is welcome to take a look in the meantime!

**Comment by francislavoie** [comment]
@mholt that's because the same site block is `:8443, *.xx.yy:8443`. It's ambiguous. How is it supposed to both match all domains, but also only match `*.xx.yy`? The adapter removes the top-level host matcher because it's redundant.

If we didn't remove the host matcher, then we'd have to have two copies of all the site's handlers in the site, which doesn't make sense. What we're doing now is the only sensible option.

**Comment by mholt** [comment]
Wait, the original post has been edited about a dozen times now. I'm trying to backtrack and see what I'm actually testing locally.

@lxhao61 If you could please try not to avoid frequent edits it would make our troubleshooting more effective.

**Comment by mholt** [comment]
Ok, so yeah, because the site block has a host-less key, it doesn't make sense to have a host matcher, thus we don't put it in a subroute.

I wonder if we should be putting the route in a subroute, though, to avoid auto-HTTPS on those domains in the host matcher within a site definition. I think we only intend to do auto-HTTPS on domains at the top of a site block.

**Comment by lxhao61** [comment]
> Wait, the original post has been edited about a dozen times now. I'm trying to backtrack and see what I'm actually testing locally since it no longer matches what OP has above.
> 
> @lxhao61 If you could please try to avoid frequent edits it would make our troubleshooting more effective.

OK I'm sorry, I wanted to express it very accurately, so I edited it, and now I will not edit it anymore.

**Comment by francislavoie** [comment]
> I wonder if we should be putting the route in a subroute, though

Yeah that seems sensible.

**Comment by mholt** [comment]
Even more relevant:

https://github.com/caddyserver/caddy/blob/b4e28af953aa02fb44f88c155956488b1f473348/caddyconfig/httpcaddyfile/httptype.go#L910-L914

**Comment by mholt** [comment]
@lxhao61 I've pushed a fix to #5130 -- could you try it out please?

**Comment by lxhao61** [comment]
> @lxhao61 I've pushed a fix to #5130 -- could you try it out please?

I tested and the problem is not solved.remove `auto_https off`, the program still automatically applies for a certificate for the subdomain.
In addition, I found that the converted JSON configuration, the non-HTTPS configuration block in the example automatically nested sub-routes, and felt that it was unnecessary to complicate.

**Comment by mholt** [comment]
@lxhao61 Hm, are you sure? Can you post the output of `caddy version` followed by `caddy adapt` for your config?

**Comment by lxhao61** [comment]
> @lxhao61 Hm, are you sure? Can you post the output of `caddy version` followed by `caddy adapt` for your config?

```
root@yisu-6310a6cff2a61:~# /usr/local/bin/caddy/caddy version
v2.6.2-0.20221007235644-0b52f7a300dd h1:s4IrV58buko8RQ4xBHRoBebRRCbA75BUqbPAFLcRcV0=
root@yisu-6310a6cff2a61:~#
root@yisu-6310a6cff2a61:~# /usr/local/bin/caddy/caddy adapt --config /usr/local/etc/caddy/Caddyfile --adapter caddyfile --pretty
{
        "admin": {
                "disabled": true
        },
        "logging": {
                "logs": {
                        "default": {
                                "writer": {
                                        "filename": "/var/log/caddy/access.log",
                                        "output": "file"
                                },
                                "level": "ERROR"
                        }
                }
        },
        "storage": {
                "module": "file_system",
                "root": "/home/tls"
        },
        "apps": {
                "http": {
                        "servers": {
                                "srv0": {
                                        "listen": [
                                                "127.0.0.1:8443"
                                        ],
                                        "listener_wrappers": [
                                                {
                                                        "wrapper": "proxy_protocol"
                                                },
                                                {
                                                        "wrapper": "tls"
                                                }
                                        ],
                                        "routes": [
                                                {
                                                        "handle": [
                                                                {
                                                                        "auth_pass_deprecated": "pass",
                                                                        "auth_user_deprecated": "user",
                                                                        "handler": "forward_proxy",
                                                                        "hide_ip": true,
                                                                        "hide_via": true,
                                                                        "probe_resistance": {}
                                                                }
                                                        ]
                                                },
                                                {
                                                        "match": [
                                                                {
                                                                        "host": [
                                                                                "wh.xx.yy",
                                                                                "ch.xx.yy"
                                                                        ]
                                                                }
                                                        ],
                                                        "handle": [
                                                                {
                                                                        "handler": "file_server",
                                                                        "hide": [
                                                                                "/usr/local/etc/caddy/Caddyfile"
                                                                        ],
                                                                        "root": "/var/www/html"
                                                                }
                                                        ]
                                                }
                                        ],
                                        "tls_connection_policies": [
                                                {
                                                        "default_sni": "*.xx.yy"
                                                }
                                        ],
                                        "automatic_https": {
                                                "disable": true
                                        }
                                },
                                "srv1": {
                                        "listen": [
                                                ":80"
                                        ],
                                        "routes": [
                                                {
                                                        "handle": [
                                                                {
                                                                        "handler": "subroute",
                                                                        "routes": [
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "static_response",
                                                                                                        "headers": {
                                                                                                                "Location": [
                                                                                                                        "https://{http.request.host}{http.request.uri}"
                                                                                                                ]
                                                                                                        },
                                                                                                        "status_code": 301
                                                                                                }
                                                                                        ]
                                                                                }
                                                                        ]
                                                                }
                                                        ],
                                                        "terminal": true
                                                }
                                        ],
                                        "automatic_https": {
                                                "disable": true
                                        }
                                },
                                "srv2": {
                                        "listen": [
                                                "unix//dev/shm/h1h2c.sock"
                                        ],
                                        "listener_wrappers": [
                                                {
                                                        "wrapper": "proxy_protocol"
                                                }
                                        ],
                                        "routes": [
                                                {
                                                        "handle": [
                                                                {
                                                                        "handler": "subroute",
                                                                        "routes": [
                                                                                {
                                                                                        "handle": [
                                                                                                {
                                                                                                        "handler": "file_server",
                                                                                                        "hide": [
                                                                                                                "/usr/local/etc/caddy/Caddyfile"
                                                                                                        ],
                                                                                                        "root": "/var/www/html"
                                                                                                }
                                                                                        ]
                                                                                }
                                                                        ]
                                                                }
                                                        ],
                                                        "terminal": true
                                                }
                                        ],
                                        "automatic_https": {
                                                "disable": true
                                        },
                                        "protocols": [
                                                "h1",
                                                "h2c"
                                        ]
                                }
                        }
                },
                "tls": {
                        "automation": {
                                "policies": [
                                        {
                                                "issuers": [
                                                        {
                                                                "challenges": {
                                                                        "dns": {
                                                                                "provider": {
                                                                                        "api_token": "b8a33cc1-1111-2222-3333-b049bf07988f",
                                                                                        "name": "duckdns",
                                                                                        "override_domain": "zz.duckdns.org"
                                                                                }
                                                                        }
                                                                },
                                                                "module": "zerossl"
                                                        }
                                                ]
                                        }
                                ]
                        }
                }
        }
}
```

**Comment by mholt** [comment]
Thanks, that's helpful. I'll look into it some more

**Comment by lxhao61** [comment]
You're welcome, I'm happy to assist.
Based on this situation, I personally suggest that canceling the `host matcher` in JSON will implicitly activate the automatic HTTPS mechanism, and the `site address` (implicitly activated automatic HTTPS) in the Caddyfile corresponds to `automate` in JSON.
This makes the `host matcher` flexible to use, and also solves the correspondence between Caddyfile and JSON automatic HTTPS.

**Comment by mholt** [comment]
@lxhao61 I appreciate that. Unfortunately that breaks the paradigm of "automatic HTTPS" though, as now users need to configure certificate management separately from their site config (by specifying their domains in "automate" rather than intuitively with their site configuration).

The result I'd be expecting you to have is for your `127.0.0.1:8443` site to have its route encapsulated by a `subroute` like your `unix//dev/shm/h1h2c.sock` and `:80` sites are, and for those sites to NOT be encapsulated by a `subroute`. So actually, the result should be backwards / the reverse from what it is now. :thinking:

**Comment by lxhao61** [comment]
@mholt Oh, keep the paradigm of automatic HTTPS. Is it possible to sink the `auto_https` parameter into `Server Options`, so as to accurately prohibit the certificate application for the host (subdomain) in the site, which is as flexible and equivalent as `automatic_https` in JSON.

**Comment by mholt** [comment]
All we have to do is get the subroute wrapping right; if it's wrapped in a subroute, auto-HTTPS won't be applied to that host matcher.

**Comment by lxhao61** [comment]
> All we have to do is get the subroute wrapping right; if it's wrapped in a subroute, auto-HTTPS won't be applied to that host matcher.

In my test, I found that replacing `auto_https off` with `auto_https disable_certs` in the above example can disable the `127.0.0.1:8443` site neutron domain name (`host matcher`) to apply for a certificate.
But if the `unix//dev/shm/h1h2c.sock` site also has `host matcher`, then something goes wrong.
Then to completely solve this kind of problem, it may be the best way to sink the `auto_https` parameter into the `Server Options`, I hope you will consider it.

**Comment by mholt** [comment]
@lxhao61 Ah, my apologies -- after reviewing, apparently an undo operation I did got undone itself (happy fingers?) and I inverted the boolean logic of the `wrapInSubroute` variable. I just pushed a commit that fixes it, and the tests which were failing on my machine (which I didn't notice, because I didn't notice I inadvertently changed the file after running the tests) are now passing.

This _should_ work for you. Can you please try that branch on its latest commit? Thanks!

**Comment by lxhao61** [comment]
@mholt It has been tested, the `subroute` method is successful, and the problem is solved: after deleting auto_https off, the subdomain will not automatically apply for a certificate. Thanks!

**Comment by mholt** [comment]
@lxhao61 Excellent!! Thank you for confirming! I will get this merged and tag 2.6.2, hopefully today. (If not, tomorrow.)
